src/rm.js

```diff
@@ -91,6 +96,56 @@ function isWriteable(file) {
   return writePermission;
 }
 
+function handleFile(file, options) {
+  if (options.force || isWriteable(file)) {
+    // -f was passed, or file is writable, so it can be removed
+    common.unlinkSync(file);
+  } else {
+    common.error('permission denied: ' + file, { continue: true });
+  }
+}
+
+function handleDirectory(file, options) {
+  if (options.recursive) {
+    // -r was passed, so directory can be removed
+    rmdirSyncRecursive(file, options.force);
+  } else {
+    common.error('path is a directory', { continue: true });
+  }
+}
+
+function handleSymbolicLink(file, options) {
+  var stats;
+  try {
+    stats = fs.statSync(file);
+  } catch (e) {
+    // symlink is bad, so just remove it
+    common.unlinkSync(file);
+    return;
+  }
+
+  if (stats.isFile()) {
+    common.unlinkSync(file);
+  } else if (stats.isDirectory()) {
+    if (file[file.length - 1] === '/') {
+      // trailing separator, so remove the contents, not the link
+      if (options.recursive) {
+        // -r was passed, so directory can be removed
+        var link = file.slice(0, -1); // remove the '/' character for readlinkSync
+        var dir = fs.readlinkSync(link);
+        var dirpath = path.resolve(path.dirname(link), dir);
+        var fromSymlink = true;
+        rmdirSyncRecursive(dirpath, options.force, fromSymlink);
+      } else {
```

Comment:
You're right, `fs.readdirSync()` works when passed symlinks, so that can be simplified. I think the `fromSymlink` logic still needs to be there, so that `rmdirSyncRecursive` knows not to delete the directory.